### PR TITLE
Rotation control and misc

### PR DESCRIPTION
**Pull request by OliverW10**
Added rotation control and changed various things, everything but the pneumatic has been tested on the robot

### CODE REVIEW

File: robot.py [hunk 1]
```diff
@@ -58,8 +58,8 @@ def createObjects(self):
         self.shooter_outer_motor = rev.CANSparkMax(2, rev.MotorType.kBrushless)
 
         self.colour_sensor = rev.color.ColorSensorV3(wpilib.I2C.Port.kOnboard)
-        self.spinner_motor = wpilib.Spark(2)
-        self.spinner_solenoid = wpilib.DoubleSolenoid(2, 3)
+        self.spinner_motor = wpilib.Spark(4)
+        self.spinner_solenoid = wpilib.Solenoid(2, 3)
```
Comment (auscompgeek) [inline robot.py]: From my understanding, we don't have a PCM with ID 2, and don't plan to… are you sure this is correct?

Comment (OliverW10) [reply robot.py]: Fixed

File: components/spinner.py [hunk 2]
```diff
@@ -3,29 +3,128 @@
 import rev.color
 
 
+class Colour:
+    def __init__(self, red: float, green: float, blue: float):
+        self.red = red
+        self.green = green
+        self.blue = blue
+
+    def dist(self, c) -> tuple:
```
Comment (auscompgeek) [inline components/spinner.py]: Is this type hint correct?

File: components/spinner.py [hunk 3]
```diff
@@ -3,29 +3,128 @@
...
-        self.spinner_solenoid.set(wpilib.DoubleSolenoid.Value.kReverse)
+        distances = {}
+        for col_name, col_value in self.WHEEL_COLOURS.items():
+            distances[col_name] = col_value.dist(self.sensed_colour)
+
+        if (
+            distances[min(distances, key=lambda col: distances[col])]
+            < self.MAX_COLOUR_DIST
+        ):
+            return min(distances, key=lambda col: distances[col])
```
Comment (auscompgeek) [inline components/spinner.py]: This would be much more readable if you only did this once outside this if statement.

File: components/spinner.py [hunk 4]
```diff
@@ -3,29 +3,128 @@
...
+
+    def go_to_colour(self, colour: str) -> bool:  # do position control
+        self.required_colour = colour
+        distance = self.get_wheel_dist()
+        self.spinner_motor.set(distance * self.POSITION_SPIN_FACTOR)
+
+        if distance == 0:
+            return True
+        else:
+            return False
```
Comment (auscompgeek) [inline components/spinner.py]: `return distance == 0` instead?

File: components/spinner.py [hunk 5]
```diff
@@ -3,29 +3,128 @@
...
+
+        if self.rotations >= 3.5:
+            self.rotations = 0
+            self.spinner_motor.set(0)
+            return True
+        else:
+            return False
+
+    def execute(self):
+        pass
```
Comment (auscompgeek) [inline components/spinner.py]: There should be logic in here to actually run the actuators.

File: components/spinner.py (hunk 3 above)
Comment (mlists) [reply components/spinner.py]: Also you remove a redundant calculation

File: components/spinner.py [hunk 6]
```diff
@@ -3,29 +3,128 @@
...
-        self.spinner_solenoid.set(wpilib.DoubleSolenoid.Value.kReverse)
+    def on_enable(self) -> None:
+        self.spinner_motor.set(0)
+        self.raise_wheel()
 
-    def execute(self):
-        pass
+    def raise_wheel(self) -> None:  # moves wheel up
+        self.spinner_solenoid.set(False)
+        self.piston_state = "up"
```
Comment (mlists) [inline components/spinner.py]: This doesn't appear to be used

File: components/spinner.py [hunk 7]
```diff
@@ -3,29 +3,128 @@
...
+            return True
+        else:
+            return False
+
+    def do_rotation(self) -> bool:
+        spin_speed = 1 - (
+            self.ROTATION_SPIN_FACTOR
+            ** (self.rotations - self.ROTATION_TARGET_ROTATIONS)
+            * self.ROTATION_MAX_SPEED
+        )
```
Comment (mlists) [inline components/spinner.py]: This does not currently ensure that the speed is below 60rpm, if you take the diameters of your wheel as constants you can use your assume the your worst case displacement on the wheel to do this calculation

File: controllers/spinner.py [hunk 8]
```diff
@@ -2,107 +2,30 @@
...
-        pass
+    def rotation_control(self):
+        self.spinner.lower_wheel()
+        self.state = "rotation"
 
     def execute(self):
-        if self.state == "position":
-            self.position()
+        if self.state == "position":  # TODO get FMS info
+            if self.spinner.go_to_colour(self.required_colour):
```
Comment (mlists) [inline controllers/spinner.py]: I don't think that this single call should be both doing things and returning data, I would recommend splitting this between two functions

File: controllers/spinner.py [hunk 9]
```diff
@@ -2,107 +2,30 @@
...
-            return 0
-        distance = (
-            self.WHEEL_ORDER.index(self.required_colour)
-            + self.COLOUR_OFFSET
-            - self.WHEEL_ORDER.index(current_colour)
-        )
-        if distance > 2:
-            return distance - 4
-        return distance
+            if self.spinner.do_rotation():
```
Comment (mlists) [inline controllers/spinner.py]: As above, would recommend making two calls

File: components/spinner.py [hunk 10]
```diff
@@ -3,29 +3,128 @@
...
+            - self.WHEEL_ORDER.index(current_colour)
+        )
+        if distance > 2:
+            return distance - 4
+        return distance
+
+    def go_to_colour(self, colour: str) -> bool:  # do position control
+        self.required_colour = colour
+        distance = self.get_wheel_dist()
+        self.spinner_motor.set(distance * self.POSITION_SPIN_FACTOR)
```
Comment (mlists) [inline components/spinner.py]: Calls like this that actually send commands to motors belong in execute. Other functions should be setting the parameters required for the motor calls. See [here](https://robotpy.readthedocs.io/en/stable/frameworks/magicbot.html#)

**Pull request closed without merging**